readme.txt
============================================
The Ordering question type for Moodle >= 2.9
============================================

The ordering question type displays several items in a random order
which the user then drags into the correct sequential order.

    * an ordering question can display all items or a subset of items

    * items can be plain text or formatted HTML, including text,
      images, audio and video.

    * items can be listed vertically or horizontally

    * several grading methods are available, ranging from a simple
      all-or-nothing grade, to more complex partial grades that
      consider the placement of each item relative to other items

============================================
To INSTALL or UPDATE this plugin
============================================

    1. get the files for this plugin from any one of the following locations:

        (a) GIT: https://github.com/gbateson/moodle-qtype_ordering.git
        (b) zip: the Moodle.org -> Plugins repository (search for Reader)
        (c) zip: http://bateson.kanazawa-gu.ac.jp/zip/plugins_qtype_ordering.zip

       If you are installing from a zip file, unzip the zip file, to create a folder
       called "ordering" and upload or move this folder into the "question/type" folder
       on your Moodle >= 2.9 site, to create a new folder at "question/type/ordering".

    2. log in to Moodle as administrator to initiate install/upgrade

        if install/upgrade does not begin automatically, you can initiate it manually
        by navigating to the following link:
        Settings -> Site administration -> Notifications

============================================
Further information
============================================

    For more information, tutorials and online discussion forums, please visit:
    http://moodlereader.org/


============================================
Development history and credits
============================================

This plugin was originally developed by Thomas Robb (Kyoto Sangyo University, Japan)
and Serafim Panov for the ordering questions in the Reader activity module for Moodle 1.x.

Moodle 2.x development was passed to Gordon Bateson (Kochi University of Technology, Japan)
funded by the Moodle Association of Japan, with partial funding from Sapporo Gakuin University.

Coding enhancements and standardization in preparation for future merging into Moodle core
were undertaken by Vadim Dvorovenko (Kemerovo State University of Culture and Arts, Russia).

Development of "relative to correct" grading method commissioned by Anatoliy Markiv and
sponsored by Kings College London, UK.

Code for privacy and automatic testing was added by Tim Hunt and Mahmoud Kassaei of Open
University, UK.

Thanks also to Dominique Bauer (ETS, University of Quebec, Canada) for his help and ideas
in testing the "Item selection type" and "Grading type", particularly the "Longest
ordered/contiguous subset" grading type.

many thanks to all concerned!


CHANGES.txt
========================================
Change log for qtype_ordering
========================================

2022-09-27 (07)
 - This version works with Moodle 4.0.
 - New option to tell students how many choices were right, as part of the automatic feedback.
 - Fixed some bugs with the layout of horizontal ordering questions.

2022-07-08 (06)
 - optimize code to ensure that recorded responses are no longer than 100 bytes.

2021-09-14 (04)
 - modify markdown syntax for long help messages (for compatability with Moodle 3.11)

2021-06-13 (03)
 - fix Issue #63. Restore missing method 'set_default_value()'

2021-06-09 (02)
 - add message to CHANGES.txt

2021-06-09 (02)
 - in edit form, rename 'get_default_value()' method to 'get_my_default_value()' for compatability with Moodle >= 3.10

2021-06-08 (01)
 - modify declaration of parameters accepted by get_default_value method in the edit form

2019-12-14 (00)
 - Thanks Tim Hunt: (1) Update Behat tests to use the MDL-66821 changes (2) Fix Behat test that was failing
 - Thanks HuongNV13: (1) Items crash together horizontally (2) Number/letter no longer tracks with dragged item

2019-12-12 (98)
 - fix JS error on Moodle >= 3.7 caused by incorrect syntax in 'define' statement in key_codes.js

2019-10-24 (96)
 - standardize layout and spelling in CHANGES.txt

2019-10-06 (94)
 - fix behavior in Adaptive mode, with 'Whether correct' review option enabled

2019-07-31 (93)
 - (github-issue #43 and #44): rename upper case numberstyles to ABCD and IIII in order to differentiate from lower case styles on DBs, such as MySQL, which use case insensitive indexes.

2019-07-12 (92)
 - rename 'answernumbering' to 'numberingstyle' and include this field in import/export of xml/gift formats.

2019-07-11 (91)
 - intermittent commit to rename answeringnumbering to numberingstyle

2019-07-11 (90)
 - Qtype_ordering: Numbering for lists (123.., abc...) (#42) Thanks Mahmoud. Great job!

2019-03-06 (89)
 - fix URL of secondary zip download site

2019-03-06 (89)
 - fix small typo in behat test file

2019-03-06 (88)
 - fix small typo in language file and optimize logic in form validation

2019-03-05 (87)
 - Qtype ordering should not allow duplicatue items (#40) Thank you Mahmoud Kassaei for your continuing work to improve the Ordering question type

2019-02-19 (86)
 - small fix to spelling and whitespace

2019-02-19 (86)
 - qtype_ordering: Ordering codechecker (#39) * Code checker issues on edit_ordering_form.php * function question_register_questiontype() not needed in questiontype.php * Code checker issues in questiontype.php * Code checker issues in question.php * Code checker issues in renderer.php * Code checker issues in version.php

2019-02-08 (85)
 - Merge pull request #38 from mkassaei/ordering_cleanup Ordering cleanup - thanks Mahmoud Kassaei!

2019-01-25 (84)
 - remove deprecated jquery folder and amd/*.js files

2019-01-25 (84)
 - Merge pull request #37 from timhunt/qtype-ordering-js-css New JavaScript implementation - many thanks  to Tim Hunt and Mahmoud Kassei of Open University

2019-01-08 (83)
 - add showgrading field to backup file

2018-12-26 (82)
 - restore accidentally removed settings.php

2018-12-26 (82)
 - (readme.txt) added credits for recent improvements: Thanks to Tim Hunt, Mahmoud Kassaei, and Dominique Bauer

2018-12-22 (81)
 - fix Github issue #36: remove contenteditable from LI element, and instead attach to ancestor element in DOM

2018-12-22 (80)
 - fix Github issue #36: draggable items become stuck to cursor when logged as student in Boost theme on Moodle 3.6

2018-12-21 (79)
 - improve logic and of code to get_subset and code to add repeated elements to form

2018-12-15 (78)
 - fix Github issue #35: Clicking Add to add items more than once without saving the question in between has incorrect and unwanted behavior

2018-12-12 (77)
 - rename variables to improve readability of loop that detects ordered subsets

2018-12-12 (77)
 - optimize code and comments in loop to detect ordered subsets
 - Merge pull request #34 from timhunt/walkthrough-unit-test Unit tests of the question as a whole - Thanks to Tim Hunt

2018-12-11 (76)
 - fix detection of longest ordered/contiguous subset grading type

2018-11-29 (75)
 - fix detection of longest ordered/contiguous subset grading type
 - fix detection of Longest ordered/contiguous subset where two or more such subsets exist
 - Merge pull request #32 from timhunt/response-analysis Response analysis - many thank Tim Hunt
 - fix detection of Longest ordered/contiguous subset where two or more such subsets exist

2018-11-26 (74)
 - fix indexing of items when "Item selection type" is "Select a contiguous subset of items". The incorrect indexing resulted in less than the exptected number of items being displayed.

2018-11-19 (73)
 - update version.php for recent BEHAT modifications

2018-11-01 (72)
 - added new setting to specify number of new draggable items to be added to a question


edit_ordering_form.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Defines the editing form for the ordering question type.
 *
 * @package    qtype_ordering
 * @copyright  2013 Gordon Bateson (gordon.bateson@gmail.com)
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

// Prevent direct access to this script.
defined('MOODLE_INTERNAL') || die();

// Include required files.
require_once($CFG->dirroot.'/question/type/ordering/question.php');

/**
 * Ordering editing form definition
 *
 * @copyright  2013 Gordon Bateson (gordon.bateson@gmail.com)
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */
class qtype_ordering_edit_form extends question_edit_form {

    /** Rows count in answer field */
    const TEXTFIELD_ROWS = 2;

    /** Cols count in answer field */
    const TEXTFIELD_COLS = 60;

    /** Number of answers in question by default */
    const NUM_ITEMS_DEFAULT = 6;

    /** Minimum number of answers to show */
    const NUM_ITEMS_MIN = 3;

    /** Number of answers to add on demand */
    const NUM_ITEMS_ADD = 1;

    /**
     * qtype is plugin name without leading "qtype_"
     */
    public function qtype() {
        return 'ordering';
    }

    /**
     * Plugin name is class name without trailing "_edit_form"
     */
    public function plugin_name() {
        return 'qtype_ordering';
    }

    /**
     * Add question-type specific form fields.
     *
     * @param object $mform the form being built.
     */
    public function definition_inner($mform) {

        // Cache this plugins name.
        $plugin = 'qtype_ordering';

        // Field for layouttype.
        $name = 'layouttype';
        $label = get_string($name, $plugin);
        $options = qtype_ordering_question::get_layout_types();
        $mform->addElement('select', $name, $label, $options);
        $mform->addHelpButton($name, $name, $plugin);
        $mform->setDefault($name, $this->get_my_default_value($name, qtype_ordering_question::LAYOUT_VERTICAL));

        // Field for selecttype.
        $name = 'selecttype';
        $label = get_string($name, $plugin);
        $options = qtype_ordering_question::get_select_types();
        $mform->addElement('select', $name, $label, $options);
        $mform->addHelpButton($name, $name, $plugin);
        $mform->setDefault($name, $this->get_my_default_value($name, qtype_ordering_question::SELECT_ALL));

        // Field for selectcount.
        $name = 'selectcount';
        $label = get_string($name, $plugin);
        $options = array(0 => get_string('all'));
        for ($i = 3; $i <= 20; $i++) {
            $options[$i] = $i;
        }
        $mform->addElement('select', $name, $label, $options);
        $mform->disabledIf($name, 'selecttype', 'eq', 0);
        $mform->addHelpButton($name, $name, $plugin);
        $mform->setDefault($name, 6);

        // Field for gradingtype.
        $name = 'gradingtype';
        $label = get_string($name, $plugin);
        $options = qtype_ordering_question::get_grading_types();
        $mform->addElement('select', $name, $label, $options);
        $mform->addHelpButton($name, $name, $plugin);
        $mform->setDefault($name, $this->get_my_default_value($name, qtype_ordering_question::GRADING_ABSOLUTE_POSITION));

        // Field for showgrading.
        $name = 'showgrading';
        $label = get_string($name, $plugin);
        $options = array(0 => get_string('hide'),
                         1 => get_string('show'));
        $mform->addElement('select', $name, $label, $options);
        $mform->addHelpButton($name, $name, $plugin);
        $mform->setDefault($name, $this->get_my_default_value($name, 1));

        $name = 'numberingstyle';
        $label = get_string($name, $plugin);
        $options = qtype_ordering_question::get_numbering_styles();
        $mform->addElement('select', $name, $label, $options);
        $mform->addHelpButton($name, $name, $plugin);
        $mform->setDefault($name, $this->get_my_default_value($name, qtype_ordering_question::NUMBERING_STYLE_DEFAULT));

        $elements = array();
        $options = array();

        $name = 'answerheader';
        $label = get_string($name, $plugin);
        $elements[] = $mform->createElement('header', $name, $label);
        $options[$name] = array('expanded' => true);

        $name = 'answer';
        $elements[] = $mform->createElement('editor', $name, $label, $this->get_editor_attributes(), $this->get_editor_options());
        $elements[] = $mform->createElement('submit', $name . 'removeeditor', get_string('removeeditor', $plugin),
                array('onclick' => 'skipClientValidation = true;'));
        $options[$name] = array('type' => PARAM_RAW);

        $this->add_repeat_elements($mform, $name, $elements, $options);

        // Adjust HTML editor and removal buttons.
        $this->adjust_html_editors($mform, $name);

        // Adding feedback fields (=Combined feedback).
        $this->add_combined_feedback_fields(true);

        // Adding interactive settings (=Multiple tries).
        $this->add_interactive_settings(false, true);
    }

    /**
     * Returns answer repeats count
     *
     * @param object $question
     * @return int
     */
    protected function get_answer_repeats($question) {
        if (isset($question->id)) {
            $repeats = count($question->options->answers);
        } else {
            $repeats = self::NUM_ITEMS_DEFAULT;
        }
        if ($repeats < self::NUM_ITEMS_MIN) {
            $repeats = self::NUM_ITEMS_MIN;
        }
        return $repeats;
    }

    /**
     * Returns editor attributes
     *
     * @return array
     */
    protected function get_editor_attributes() {
        return array(
            'rows'  => self::TEXTFIELD_ROWS,
            'cols'  => self::TEXTFIELD_COLS
        );
    }

    /**
     * Returns editor options
     *
     * @return array
     */
    protected function get_editor_options() {
        return array(
            'context'  => $this->context,
            'maxfiles' => EDITOR_UNLIMITED_FILES,
            'noclean'  => true
        );
    }

    /**
     * Resets editor format to specified
     *
     * @param object $editor
     * @param int $format
     * @return int
     */
    protected function reset_editor_format($editor, $format=FORMAT_MOODLE) {
        $value = $editor->getValue();
        $value['format'] = $format;
        $value = $editor->setValue($value);
        return $format;
    }

    /**
     * Adjust HTML editor and removal buttons.
     *
     * @param object $mform
     * @param string $name
     * @param int $repeats
     */
    protected function adjust_html_editors($mform, $name) {

        // Cache the number of formats supported
        // by the preferred editor for each format.
        $count = array();

        if (isset($this->question->options->answers)) {
            $ids = array_keys($this->question->options->answers);
        } else {
            $ids = array();
        }

        $defaultanswerformat = get_config('qtype_ordering', 'defaultanswerformat');

        $repeats = 'count'.$name.'s'; // E.g. countanswers.
        if ($mform->elementExists($repeats)) {
            // Use mform element to get number of repeats.
            $repeats = $mform->getElement($repeats)->getValue();
        } else {
            // Determine number of repeats by object sniffing.
            $repeats = 0;
            while ($mform->elementExists($name."[$repeats]")) {
                $repeats++;
            }
        }

        for ($i = 0; $i < $repeats; $i++) {
            $editor = $mform->getElement($name."[$i]");

            if (isset($ids[$i])) {
                $id = $ids[$i];
            } else {
                $id = 0;
            }

            // The old/new name of the button to remove the HTML editor
            // old : the name of the button when added by repeat_elements
            // new : the simplified name of the button to satisfy "no_submit_button_pressed()" in lib/formslib.php.
            $oldname = $name.'removeeditor['.$i.']';
            $newname = $name.'removeeditor_'.$i;

            // Remove HTML editor, if necessary.
            if (optional_param($newname, 0, PARAM_RAW)) {
                $format = $this->reset_editor_format($editor, FORMAT_MOODLE);
                $_POST['answer'][$i]['format'] = $format; // Overwrite incoming data.
            } else if ($id) {
                $format = $this->question->options->answers[$id]->answerformat;
            } else {
                $format = $this->reset_editor_format($editor, $defaultanswerformat);
            }

            // Check we have a submit button - it should always be there !!
            if ($mform->elementExists($oldname)) {
                if (! isset($count[$format])) {
                    $editor = editors_get_preferred_editor($format);
                    $count[$format] = $editor->get_supported_formats();
                    $count[$format] = count($count[$format]);
                }
                if ($count[$format] > 1) {
                    $mform->removeElement($oldname);
                } else {
                    $submit = $mform->getElement($oldname);
                    $submit->setName($newname);
                }
                $mform->registerNoSubmitButton($newname);
            }
        }
    }

    /**
     * Create the form elements required by one hint.
     *
     * @param string $withclearwrong Whether this quesiton type uses the 'Clear wrong' option on hints.
     * @param string $withshownumpartscorrect Whether this quesiton type uses the 'Show num parts correct' option on hints.
     * @return array Form field elements for one hint.
     */
    protected function get_hint_fields($withclearwrong = false, $withshownumpartscorrect = false) {
        $mform = $this->_form;

        $repeated = [];
        $repeated[] = $mform->createElement('editor', 'hint', get_string('hintn', 'question'),
            ['rows' => 5], $this->editoroptions);
        $repeatedoptions['hint']['type'] = PARAM_RAW;

        $optionelements = [];

        if ($withshownumpartscorrect) {
            $optionelements[] = $mform->createElement('advcheckbox', 'hintshownumcorrect', '',
                get_string('shownumpartscorrect', 'question'));
        }

        $optionelements[] = $mform->createElement('advcheckbox', 'hintoptions', '',
            get_string('highlightresponse', 'qtype_ordering'));

        if (count($optionelements)) {
            $repeated[] = $mform->createElement('group', 'hintoptions',
                get_string('hintnoptions', 'question'), $optionelements, null, false);
        }

        return [$repeated, $repeatedoptions];
    }

    /**
     * Perform an preprocessing needed on the data passed to {@link set_data()}
     * before it is used to initialise the form.
     * @param object $question the data being passed to the form.
     * @return object $question the modified data.
     */
    public function data_preprocessing($question) {

        $question = parent::data_preprocessing($question);
        $question = $this->data_preprocessing_answers($question, true);

        // Preprocess feedback.
        $question = $this->data_preprocessing_combined_feedback($question, true);

        $question = $this->data_preprocessing_hints($question, false, true);

        // Preprocess answers and fractions.
        $question->answer     = array();
        $question->fraction   = array();

        if (empty($question->options->answers)) {
            $answerids = array();
        } else {
            $answerids = array_keys($question->options->answers);
        }

        $defaultanswerformat = get_config('qtype_ordering', 'defaultanswerformat');
        $repeats = $this->get_answer_repeats($question);
        for ($i = 0; $i < $repeats; $i++) {

            if ($answerid = array_shift($answerids)) {
                $answer = $question->options->answers[$answerid];
            } else {
                $answer = (object)array('answer' => '',
                                        'answerformat' => $defaultanswerformat);
                $answerid = 0;
            }

            if (empty($question->id)) {
                $question->answer[$i] = $answer->answer;
            } else {
                $itemid = file_get_submitted_draft_itemid("answer[$i]");
                $format = $answer->answerformat;
                $text = file_prepare_draft_area($itemid, $this->context->id, 'question', 'answer',
                                                $answerid, $this->editoroptions, $answer->answer);
                $question->answer[$i] = array('text' => $text,
                                              'format' => $format,
                                              'itemid' => $itemid);
            }
            $question->fraction[$i] = ($i + 1);
        }

        // Defining default values.
        $names = array(
            'layouttype'  => qtype_ordering_question::LAYOUT_VERTICAL,
            'selecttype'  => qtype_ordering_question::SELECT_ALL,
            'selectcount' => 0, // 0 means ALL.
            'gradingtype' => qtype_ordering_question::GRADING_ABSOLUTE_POSITION,
            'showgrading' => 1,  // 1 means SHOW.
            'numberingstyle' => qtype_ordering_question::NUMBERING_STYLE_DEFAULT
        );
        foreach ($names as $name => $default) {
            if (isset($question->options->$name)) {
                $question->$name = $question->options->$name;
            } else {
                $question->$name = $this->get_my_default_value($name, $default);
            }
        }

        return $question;
    }

    /**
     * Perform the necessary preprocessing for the hint fields.
     *
     * @param object $question The data being passed to the form.
     * @param bool $withclearwrong Clear wrong hints.
     * @param bool $withshownumpartscorrect Show number correct.
     * @return object The modified data.
     */
    protected function data_preprocessing_hints($question, $withclearwrong = false, $withshownumpartscorrect = false) {
        if (empty($question->hints)) {
            return $question;
        }
        parent::data_preprocessing_hints($question, $withclearwrong, $withshownumpartscorrect);

        $question->hintoptions = [];
        foreach ($question->hints as $hint) {
            $question->hintoptions[] = $hint->options;
        }

        return $question;
    }

    /**
     * Form validation
     *
     * @param array $data array of ("fieldname"=>value) of submitted data
     * @param array $files array of uploaded files "element_name"=>tmp_file_path
     * @return array of "element_name"=>"error_description" if there are errors,
     *         or an empty array if everything is OK (true allowed for backwards compatibility too).
     */
    public function validation($data, $files) {
        $errors = array();
        $plugin = 'qtype_ordering';

        // Identify duplicates and report as an error.
        $answers = [];
        $answercount = 0;
        foreach ($data['answer'] as $answer) {
            if (is_array($answer)) {
                $answer = $answer['text'];
            }
            if ($answer = trim($answer)) {
                if (in_array($answer, $answers)) {
                    $i = array_search($answer, $answers);
                    $item = get_string('answerheader', $plugin);
                    $item = str_replace('{no}', $i + 1, $item);
                    $item = html_writer::link("#id_answerheader_$i", $item);
                    $a = (object)array('text' => $answer, 'item' => $item);
                    $errors["answer[$answercount]"] =  get_string('duplicatesnotallowed', $plugin, $a);
                } else {
                    $answers[] = $answer;
                }
                $answercount++;
            }
        }

        switch ($answercount) {
            case 0: $errors['answer[0]'] = get_string('notenoughanswers', $plugin, 2);
            case 1: $errors['answer[1]'] = get_string('notenoughanswers', $plugin, 2);
        }

        // If adding a new ordering question, update defaults.
        if (empty($errors) && empty($data['id'])) {
            $fields = array('layouttype', 'selecttype', 'selectcount',
                            'gradingtype', 'showgrading', 'numberingstyle');
            foreach ($fields as $field) {
                if (array_key_exists($field, $data)) {
                    $this->set_my_default_value($field, $data[$field]);
                }
            }
        }

        return $errors;
    }

    /**
     * Given a preference item name, returns the full
     * preference name of that item for this plugin
     *
     * @param string $name Item name
     * @return string full preference name
     */
    protected function get_my_preference_name($name) {
        return $this->plugin_name()."_$name";
    }

    /**
     * Saves default value for item
     *
     * @param string $name Item name
     * @param string|mixed|null $value
     * @return boolean (usually TRUE, unless there is an error) 
     */
    protected function set_my_default_value($name, $value) {
        if (method_exists($this, 'set_default_value')) {
            // This method doesn't exist yet, but it might one day ;-)
            return $this->set_default_value($name, $value);
        } else {
            // Until at least Moodle <= 4.0, we expect to come this way
            $name = $this->get_my_preference_name($name);
            return set_user_preferences(array($name => $value));
        }
    }

    /**
     * Returns default value for item
     *
     * @param string $name Item name
     * @param string|mixed|null $default Default value (optional, default = null)
     * @return string|mixed|null Default value for field with this $name
     */
    protected function get_my_default_value($name, $default) {
        if (method_exists($this, 'get_default_value')) {
            // Moodle >= 3.10
            return $this->get_default_value($name, $default);
        } else {
            // Moodle <= 3.9
            $name = $this->get_my_preference_name($name);
            return get_user_preferences($name, $default);
        }
    }

    /**
     * Get array of countable item types
     *
     * @return array(type => description)
     */
    protected function get_addcount_options($type, $max=10) {

        // Cache plugin name.
        $plugin = $this->plugin_name();

        // Generate options.
        $options = array();
        for ($i = 1; $i <= $max; $i++) {
            if ($i == 1) {
                $options[$i] = get_string('addsingle'.$type, $plugin);
            } else {
                $options[$i] = get_string('addmultiple'.$type.'s', $plugin, $i);
            }
        }
        return $options;
    }

    /**
     * Add repeated elements with a button allowing a selectable number of new elements
     *
     * @param object $mform the Moodle form object
     * @return voide, but will update $mform
     */
    protected function add_repeat_elements($mform, $type, $elements, $options) {

        // Cache plugin name.
        $plugin = $this->plugin_name();

        // Cache element names.
        $types = $type.'s';
        $addtypes = 'add'.$types;
        $counttypes = 'count'.$types;
        $addtypescount = $addtypes.'count';
        $addtypesgroup = $addtypes.'group';

        $repeats = $this->get_answer_repeats($this->question);

        $count = optional_param($addtypescount, self::NUM_ITEMS_ADD, PARAM_INT);

        $label = ($count == 1 ? 'addsingle'.$type : 'addmultiple'.$types);
        $label = get_string($label, $plugin, $count);

        $this->repeat_elements($elements, $repeats, $options, $counttypes, $addtypes, $count, $label, true);

        // Remove the original "Add xxx" button ...
        $mform->removeElement($addtypes);

        // ... and replace it with "Add" button + select group.
        $options = $this->get_addcount_options($type);
        $mform->addGroup(array(
            $mform->createElement('submit', $addtypes, get_string('add')),
            $mform->createElement('select', $addtypescount, '', $options)
        ), $addtypesgroup, '', ' ', false);

        // Set default value and type of select element.
        $mform->setDefault($addtypescount, $count);
        $mform->setType($addtypescount, PARAM_INT);
    }
}
